app/models/job.rb

```diff
@@ -218,4 +181,41 @@ def validate_batch_script
       @errors.add(:batch_script, batch_script.errors.full_messages)
     end
   end
+
+  def running_tasks
+    if job_type == 'ARRAY_JOB'
+      FlightScheduler.app.job_registry.running_tasks_for(self)
+    else
+      nil
+    end
+  end
+
+  def task_generator
+    if job_type == 'ARRAY_JOB'
+      @task_generator ||= FlightScheduler::ArrayTaskGenerator.new(self)
+    else
+      nil
+    end
+  end
+
+  def terminal_state?
+    TERMINAL_STATES.include?(state)
+  end
+
+  def update_array_job_state
+    return unless job_type == 'ARRAY_JOB'
+    return unless task_generator.finished? || state == 'CANCELLING'
+
+    tasks = FlightScheduler.app.job_registry.tasks_for(self)
+    return unless tasks.all?(&:terminal_state?)
+
+    if tasks.any? { |t| t.state == 'FAILED' }
+      self.state = 'FAILED'
+    elsif tasks.any? { |t| t.state == 'CANCELLED' }
+      self.state = 'CANCELLED'
+    else
+      # They must all be completed then.
+      self.state = 'COMPLETED'
+    end
```

Comment:
This conditional inherently assumes the only three terminal states are `FAILED`, `CANCELLED`, and `COMPLETED`. This probably fine for the time being, however we will need to remember to update it if and when a new terminal state is added.

An alternative is to preform something along these lines:
```
states  = task.map(&:state).uniq
self.state = TERMINAL_STATES.first { |s| states.include?(s) }
```

This assumes that `TERMINAL_STATES` in precedence order, but that isn't to hard to maintain.

Reply:
That sounds very implicit, and the kind of thing that is easy to get wrong if/when we refactor to use a state machine to handle the job's state.  I'll add a comment above `TERMINAL_STATES` about keeping the two in sync.